src/dtx/dtx_common.c

```diff
@@ -543,6 +480,214 @@ dtx_conflict(daos_handle_t coh, struct dtx_handle *dth, uuid_t po_uuid,
+			     dth->dth_intent == DAOS_INTENT_PUNCH ?
+			     true : false);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": Fail to add DTX "DF_DTI" to CoS cache: %d. "
+			"Try to commit it sychronously.\n",
+			DP_UUID(cont->sc_uuid), DP_DTI(&dth->dth_xid), rc);
+		dth->dth_sync = 1;
+	}
+
+	if (dth->dth_sync) {
+		rc = dtx_commit(cont_hdl->sch_pool->spc_uuid,
+				cont->sc_uuid, &dth->dth_dte, 1,
+				cont_hdl->sch_pool->spc_map_version);
+		if (rc != 0) {
+			D_ERROR(DF_UUID": Fail to sync commit DTX "DF_DTI
+				": rc = %d\n", DP_UUID(cont->sc_uuid),
+				DP_DTI(&dth->dth_xid), rc);
+			D_GOTO(fail, result = rc);
+		}
+	}
+
+fail:
+	if (result < 0)
+		dtx_abort(cont_hdl->sch_pool->spc_uuid, cont->sc_uuid,
+			  &dth->dth_dte, 1,
+			  cont_hdl->sch_pool->spc_map_version);
+out_free:
+	D_DEBUG(DB_TRACE,
+		"Stop the DTX "DF_DTI" ver %u, dkey %llu, intent %s, "
+		"%s, %s: rc = %d\n",
+		DP_DTI(&dth->dth_xid), dth->dth_ver,
+		(unsigned long long)dth->dth_dkey_hash,
+		dth->dth_intent == DAOS_INTENT_PUNCH ? "Punch" : "Update",
+		dth->dth_sync ? "sync" : "async",
+		dth->dth_non_rep ? "non-replicated" : "replicated", result);
+
+	if (dth->dth_dti_cos != NULL)
+		D_FREE(dth->dth_dti_cos);
+out:
+	return result > 0 ? 0 : result;
+}
+
+/**
+ * Prepare the DTX handle in DRAM.
+ *
+ * XXX: Currently, we only support to prepare the DTX against single DAOS
+ *	object and single dkey.
+ *
+ * \param dti		[IN]	The DTX identifier.
+ * \param oid		[IN]	The target object (shard) ID.
+ * \param coh		[IN]	Container open handle.
+ * \param epoch		[IN]	Epoch for the DTX.
+ * \param dkey_hash	[IN]	Hash of the dkey to be modified if applicable.
+ * \param conflict	[IN]	Hash of the dkey to be modified if applicable.
+ * \param dti_cos	[IN,OUT]The DTX array to be committed because of shared.
+ * \param dti_cos_count [IN,OUT]The @dti_cos array size.
+ * \param pm_ver	[IN]	Pool map version for the DTX.
+ * \param intent	[IN]	The intent of related modification.
+ * \param leader	[IN]	The target (to be modified) is leader or not.
+ * \param dth		[OUT]	Pointer to the DTX handle.
+ *
+ * \return			Zero on success, negative value if error.
+ */
+int
+dtx_begin(struct dtx_id *dti, daos_unit_oid_t *oid, daos_handle_t coh,
+	  daos_epoch_t epoch, uint64_t dkey_hash,
+	  struct dtx_conflict_entry *conflict, struct dtx_id *dti_cos,
+	  int dti_cos_cnt, uint32_t pm_ver, uint32_t intent,
+	  struct dtx_handle *dth)
+{
+	dtx_handle_init(dti, oid, coh, epoch, dkey_hash, pm_ver, intent,
+			conflict, dti_cos, dti_cos_cnt, false, dth);
+
+	D_DEBUG(DB_TRACE, "Start the DTX "DF_DTI" for object "DF_OID
+		" ver %u, dkey %llu, dti_cos_count %d, intent %s\n",
+		DP_DTI(&dth->dth_xid), DP_OID(oid->id_pub), dth->dth_ver,
+		(unsigned long long)dth->dth_dkey_hash, dti_cos_cnt,
+		intent == DAOS_INTENT_PUNCH ? "Punch" : "Update");
+
+	return 0;
+}
+
+int
+dtx_end(struct dtx_handle *dth, struct ds_cont_hdl *cont_hdl,
+	struct ds_cont_child *cont, struct dtx_id *dti_cos, int dti_cos_cnt,
+	int result)
```

Comment:
@dti_cos and @dti_cos_cnt are members of @dth, no need be as dtx_end() parameters.

Reply:
ah, right